[WEEK 2] 자바 사다리게임 구현

JakePark929 opened the pull request:
자바 사다리게임 구현 2주차 입니다.

참가자 이름을 `GameAttendees` 객체에 입력 받아 사다리 게임을 진행합니다.
입력 받은 이름의 길이와 참가자 수 확인 등에 대한 로직의 테스트 코드도 추가하였습니다.

바쁘시겠지만 리뷰 잘 부탁드립니다!

JakePark929 committed:
[WEEK 1] Solve ladder game

JakePark929 committed:
refactor: Optional 객체 처리 및 LadderView 공백 추가

JakePark929 committed:
[WEEK 2] Solve ladder game with test

birdieHyun commented on src/main/java/ladder/view/LadderInputView.java:
```diff
@@ -0,0 +1,25 @@
+
+import java.util.Scanner;
+
+public record LadderInputView(
+        String gameAttendees,
+        int ladderHeight
+) {
```
View 객체가 상태를 가지고 있어야 할까요?

birdieHyun commented on src/test/java/ladder/util/StringUtilTest.java:
```diff
@@ -0,0 +1,37 @@
+        assertThat(result.length()).isEqualTo(totalLength);
+        assertThat(result).isEqualTo(expected);
+    }
+}
```
첫 번째 assertThat()이 실패하게 된다면 두 번째 검증은 어떻게 될까요?

birdieHyun commented on src/main/java/ladder/domain/RandomLineStrategy.java:
```diff
@@ -0,0 +1,12 @@
+
+import java.util.Random;
+
+public class RandomLineStrategy implements HorizontalLineStrategy {
```
HorizontalLineStrategy를 implements 했지만,RandomLineStrategy는 Horizontal과 관련이 없는 객체처럼 보일 수 있지 않을까요? 네이밍을 조금 더 구체적으로 해도 좋을 것 같아요

birdieHyun commented on src/main/java/ladder/domain/Point.java:
```diff
@@ -0,0 +1,38 @@
+
+    public void setNextMovePoint(final Point nextMovePoint) {
+        this.nextMovePoint = nextMovePoint;
+    }
```
이런 메서드는 setter를 사용할 수 밖에 없지만, 일반적인 컨벤션 상 setter보단 다른 메서드명을 지어주는건 어떨까요?

birdieHyun commented on src/main/java/ladder/view/LadderInputView.java:
```diff
@@ -0,0 +1,25 @@
+
+public record LadderInputView(
+        String gameAttendees,
+        int ladderHeight
```
ladderHeight에 int 범위를 넘어서는 숫자를 입력받는다면 어떻게 될까요?

birdieHyun reviewed (approved):
고생많으셨습니다 :) 
너무 잘 작성해주셔서 크게 리뷰해드릴게 없네요 ㅜㅜ 

몇가지 적은거 참고만 해주세요!

JakePark929 committed:
refactor: LadderInputView 상태값 제거와 높이 입력 수정, 테스트코드 수정, 이름 변경

- `LadderInputView` 객체의 상태 값은 `LadderConfiguration` 객체에 위임하고 View 객체의 역할에 집중하도록 수정하였습니다.
- `StringUtilTest`에서 `fixStringCenterWithValidInput` 테스트의 input 이 비어있을 때와 5글자 이상일 때의 케이스를 추가하고 해당 케이스에 맞도록 테스트 코드를 추가하였습니다.
- `RandomLineStrategy`의 객체명을 `RandomHorizontalLineStrategy` 로 변경하여 구현체의 역할을 확실히 인지할 수 있도록 하였습니다.
- `Point` 객체의 `setNextMovePoint`메소드명을 `assignNextMovePoint`로 변경하여 메소드의 역할을 잘 인지할 수 있도록 하였습니다.
- `LadderInputView` 에서 사다리 높이를 입력 받을 때 input 값이 int 가 아니거나 음수인 경우 재입력 하도록 수정하였습니다.

JakePark929 commented:
바쁘신 와중에 친절히 리뷰 해주셔서 감사합니다.
리뷰 주신 내용을 기준으로 커밋 완료 했습니다.

서영님 리뷰 완료되면 머지하고 다음 과제 진행하도록 하겠습니다. :D

birdieHyun commented on src/test/java/ladder/util/StringUtilTest.java:
```diff
@@ -21,17 +21,27 @@ void failIfInputSizeBiggerThanBoundary(String input) {
     void fixStringCenterWithValidInput(String input, String expected) {
         int totalLength = 5;
-        String result = StringUtil.fixStringCenter(input);
 
-        assertThat(result.length()).isEqualTo(totalLength);
-        assertThat(result).isEqualTo(expected);
+        if (expected.isEmpty()) {
+            // 예외 발생을 검증
+            assertThatThrownBy(() -> StringUtil.fixStringCenter(input))
+                    .isInstanceOf(IllegalArgumentException.class)
+                    .hasMessageContaining("최대 5 글자 까지 입력 가능합니다.");
+        } else {
+            String result = StringUtil.fixStringCenter(input);
+
+            assertThat(result.length()).isEqualTo(totalLength);
+            assertThat(result).isEqualTo(expected);
+        }
     }
```
테스트 코드 내에서 분기를 나누는 것은 일반적으로 안티패턴으로 간주됩니다! 

assertAll() 키워드 학습 추천드릴게요 :)

JakePark929 committed:
refactor: StringUtilTest 테스트 코드 리뷰 반영

- `StringUtilTest` 에서 테스트 코드 내 분기를 나누는 것은 안티 패턴으로, 테스트를 나누어 진행. assertAll() 메소드 사용.

JakePark929 commented:
시험기간이라 조금 바빠져서 반영이 느렸습니다 ㅠ
리뷰 해주신 내용 반영하였고 서영님 어프루브 후에 머지하도록 하겠습니다!

BellaChooi reviewed (approved):
리뷰가 늦어지기까지 했는데 너무 잘하셔서 제가 첨언할 게 없어 죄송합니닷..ㅎ 고생하셨습니다!

JakePark929 committed:
Merge branch 'jakepark' into week2-jake
